### ISSUE TITLE
useLocalSearchParams() is always returning {} when used inside of <Stack.Screen>

### ISSUE THREAD

**Issue by petja** [expo/expo#30625]
### Minimal reproducible example

https://github.com/petja/expo-use-local-search-params-issue-repro

### Which package manager are you using? (Yarn is recommended)

pnpm

### If the issue is web-related, please select the bundler (`web.bundler` in the `app.json`)

metro

### Summary

`useLocalSearchParams()` is always returning `{}` when used inside of `<Stack.Screen>`'s `options.headerRight`. Instead of an empty object, it should return search params for the screen if given.

This worked before I upgraded from Expo 50 to 51, but not anymore.

### Environment

  expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 14.5
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 20.12.1 - ~/.nvm/versions/node/v20.12.1/bin/node
      npm: 10.6.0 - ~/.nvm/versions/node/v20.12.1/bin/npm
    Managers:
      CocoaPods: 1.15.2 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.5, iOS 17.5, macOS 14.5, tvOS 17.5, visionOS 1.2, watchOS 10.5
    IDEs:
      Android Studio: 2023.2 AI-232.10300.40.2321.11567975
      Xcode: 15.4/15F31d - /usr/bin/xcodebuild
    npmPackages:
      expo: ~51.0.22 => 51.0.22
      expo-router: ~3.5.18 => 3.5.18
      react: 18.2.0 => 18.2.0
      react-dom: 18.2.0 => 18.2.0
      react-native: 0.74.3 => 0.74.3
      react-native-web: ~0.19.10 => 0.19.12
    npmGlobalPackages:
      eas-cli: 10.2.1
    Expo Workflow: managed

**Comment by pkfms** [comment]
this worked on 3.5.17, I just updated to 3.5.20 and useLocalSearchParams() in headerTitle started returning an empty object

**Comment by petja** [comment]
I can confirm that downgrading to 3.5.17 helps. So the problem is for >=3.5.18.

**Comment by petja** [comment]
Can we somehow prioritize this issue? As I just tested, it's still present in expo-router 3.5.23. Every time I run `expo install --fix` to upgrade my other Expo packages, it also installs this broken version of expo-router, and I have to install an older version manually.

**Comment by Kaarelmm** [comment]
This is my TEMPORARY plug-in workaround for using useLocalSearchParams while staying compatible with SDK52. From my testing, I can confirm that it appears that the last working expo-router version was 3.5.17. 
I'll use this temporary workaround until the issue is resolved or I find a proper fix.

```
import { ParamListBase, useNavigation } from '@react-navigation/native'
import { StackNavigationProp } from '@react-navigation/stack'
import { useMemo } from 'react'

export const useLocalSearchParams2 = <T = undefined,>(): T => {
    const navigation = useNavigation<StackNavigationProp<ParamListBase>>()
    const navState = navigation.getState()
    const params = navState.routes?.[navState.index]?.params

    return useMemo(() => {
        if (isObjectMatchingType<T>(params)) {
            return params as T
        }
        return createUndefinedObject<T>()
    }, [params])
}

function isObjectMatchingType<T>(obj: unknown): obj is T {
    if (typeof obj !== 'object' || obj === null) return false

    return true
}

function createUndefinedObject<T>(): T {
    return new Proxy(
        {},
        {
            get: () => undefined,
        }
    ) as T
}

export default useLocalSearchParams2

```

**Comment by HenriLemos** [comment]
> I can confirm that downgrading to 3.5.17 helps. So the problem is for >=3.5.18.
You literally saved my life on college's project, thanks man <3

**Comment by petja** [comment]
I tried @Kaarelmm approach and it works with the first screen in the stack, but if I navigate to next screen in the stack, the first screen gets params of second screen which is not wanted.

**Comment by Contresito** [comment]
This is happening to me too, on version 4.0.16 of expo-router
I'm basically forced to use useGlobalSearchParams on everything for the time being, which is probably not good for performance...

**Comment by anthonysgro** [comment]
this needs to be fixed, this is still an issue on version 4.0.17 that comes bundled on the latest `create-expo-app`

**Comment by pkfms** [comment]
As a workaround I did

`import { useRoute } from '@react-navigation/native';
  
  const { params } = useRoute();`

**Comment by moritzlang** [comment]
I also experience issues with `useLocalSearchParams`. When opening a screen directly via an uri, `useLocalSearchParams` returns an empty object. Whereas `useGlobalSearchParams` returns values, but stays empty when navigating to a screen.

**Comment by FTCHD** [comment]
`useLocalSearchParams` is currently unusable in expo router v4.

`useGlobalSearchParams` resets the params to undefined even when navigating to a nested route, re-rendering the parent screen for no reason.

In short neither solutions work at the moment, making param passing in expo router impossible to do correctly.

**Comment by hoanghuy27297** [comment]
I got into the same issue, tried `success/[screen].tsx`, and didn't work as the `useLocalSearchParams` return {}. Then, I changed the route to `success/[id].tsx` and now the id has value. So, it may be a strict naming convention to make the dynamic route work in Expo router v4

**Comment by FTCHD** [comment]
any news on prioritizing this @marklawlor? params are essentially unusable at the moment.

**Comment by FTCHD** [comment]
it is very much an active issue

**Comment by softov** [comment]
I have a similar problem.

But I'm using Drawer

Inside the Drawer, I have

/(drawer)/client_list
/(drawer)/contract_list

Sometimes, from the cliente, I navigate to "contract_list?client_id=15"

Its the contract_list with some filter info.

But, when going in the drawer and navigating to "contract_list" (no params)

const params = useLocalSearchParams();
// params: {client_id: 15}

How can I navigate using the drawer, but without previous params?

**Comment by FTCHD** [comment]
not stale

**Comment by TylerJackson** [comment]
this is still an issue any update on this? useLocalSearchParams and useGlobalSearchParams only show the variables in the the top most stack navigator. This is a major bug in how you would intuitively think grabbing route params would work

**Comment by FTCHD** [comment]
My custom hook:

```
export function useSearchParams<T extends Record<string, string>>() {
    //! `useLocalSearchParams` only works on the initial render (when going to /home)
    //! so we have to update this manually using GlobalSearchParams
    //! if we use GlobalSearchParams directly, it will clear the params mid way through the navigation
    //! (when going from vector/home to vector/records or from records to [recordId] to another page for example, it shows the loading state while the navigation is still happening
    //! or worse, it removes the param making it undefined until you navigate back, making the screen load every time you go back)
    const globalSearchParams = useGlobalSearchParams<T>()

    const [_searchParams, _setSearchParams] = useState<T>({} as unknown as T)

    useFocusEffect(
        useCallback(() => {
            if (globalSearchParams === undefined) return
            _setSearchParams(globalSearchParams)
        }, [globalSearchParams])
    )

    return _searchParams
}
```

This is what I'm doing to solve it, it's stupid but it works (+ I've seen worse) 😂

**Comment by kitten** [comment]
This is a few issues mixed into one, so it's become hard to re-address this, but I'll try to pick this apart.

Regarding `useLocalSearchParams` in `headerRight`; The `headerRight` element is "above" the `Route` structure, and with inheriting react-navigation's structure but also providing per-route params driven by the file routing, the local screen params basically end up defined on the route, and hence defined below the `headerRight`, i.e. only in the route itself.

This is less of an issue with the per-route composition API now, since it's now possible to configure many per-route options from the screen itself, which means `useLocalSearchParams` will return the correct params there: https://docs.expo.dev/router/advanced/stack/#set-screen-options-dynamically

The problem with allowing layouts to contribute to this is that `useLocalSearchParams` would stop returning "local" results. Meaning, if we wanted to type this or have route-specific options in the future, that wouldn't be possible anymore, since `headerRight` could be coming from and for any screen.

Technically, at the "old" definition site of `headerRight` above a `Route`, there isn't a single shape of params, so you'll have to use `useGlobalSearchParams` instead, since there's no definition of a per-route params object outside of a singular route.

Again: This should be more intuitive now with the per-screen composition API, allowing `Stack` options to be updated from the screen itself.

---

tl;dr: `useLocalSearchParams` needs to be used in a route and is scoped to a single route's params, since params can compose downwards from multiple per-route params, which aren't "local"
